File: source/cpp/glasses.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>

#include <algorithm>
#include <numeric>


//-----------------------------------------------------------------------------
struct Glass{ 
    int64_t volume;
    int64_t capacity;
};
using Glasses = std::vector<Glass>;

std::deque<Glasses> work_queue;
std::map<decltype(Glass::volume),size_t> volumes;
std::map<decltype(Glass::volume),size_t> states;
//-----------------------------------------------------------------------------
Glass CreateGlass(int cap)
{ return {cap,cap}; }

//-----------------------------------------------------------------------------
inline Glasses empty_glass(const Glasses& state, size_t index)
{ 
  if ( state.size() < 3){
    return {};
  }
  if(state[index].volume != 0){
    Glasses new_state = state;
    new_state[index].volume = 0;
    if (new_state[0].volume | new_state[1].volume | new_state[2].volume) {
      return new_state;  
    } else {
      return {};
    }
  } 
  return {};
}

//-----------------------------------------------------------------------------
inline Glasses glass_transfer(const Glasses& state, size_t from, size_t to){
    if ( state.size() < 3){
      return {};
    }
    if ( state[from].volume != 0 && state[to].capacity != state[to].volume ){
        Glasses new_state = state;

        int64_t vaccent = state[to].capacity - state[to].volume;
        int64_t avaliable = state[from].volume;
        if ( vaccent >= avaliable){
          new_state[from].volume = 0;
          new_state[to].volume+= avaliable;
        } else {
          avaliable -= vaccent;
          new_state[from].volume -= vaccent;
          new_state[to].volume   += vaccent;
        }
        return new_state;
    }
    return {};
}
//-----------------------------------------------------------------------------
inline size_t glass_key(const Glasses& state){
  if(state.size() == 3)
  {
    return state[0].volume << 32 | state[1].volume << 16 | state[2].volume;
  } else { 
    return 0;
  }
}
//-----------------------------------------------------------------------------
int threeGlasses(std::vector<int64_t> cap) {
    volumes.clear();
    work_queue.clear();
    states.clear();
    int64_t ans = 1;
    std::sort(cap.begin(),cap.end());
    if( cap.front() == 1 ){
      ans = std::accumulate(cap.begin(), cap.end(),0);
    } else {
      volumes[std::accumulate(cap.begin(), cap.end(),0)]++;

      work_queue.push_back({ {    0 ,cap[0]}, {cap[1],cap[1]}, {cap[2],cap[2]} });
      work_queue.push_back({ {cap[0],cap[0]}, {    0 ,cap[1]}, {cap[2],cap[2]} });
      work_queue.push_back({ {cap[0],cap[0]}, {cap[1],cap[1]}, {    0 ,cap[2]} });
      
      states[glass_key({ {cap[0],cap[0]}, {cap[1],cap[1]}, {cap[2],cap[2]} })]++;

      Glasses new_state;
      while ( work_queue.begin() != work_queue.end())
      {
        Glasses& current = work_queue.front();
        if ( current.empty() )
        { 
           work_queue.pop_front();
           continue;  
        }
        
        volumes[std::accumulate(current.begin(), current.end(),0, 
          [](int64_t sum,const Glass& g){return sum + g.volume;})  ]++;
        
        Glasses new_state     = glass_transfer(current,0,1);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        new_state   = glass_transfer(current,0,2);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        new_state     = glass_transfer(current,1,0);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        new_state   = glass_transfer(current,1,2);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        new_state   = glass_transfer(current,2,0);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        new_state   = glass_transfer(current,2,1);
        if( !states[glass_key(new_state)]++ )
        {
          work_queue.push_back( new_state );
          work_queue.push_back( empty_glass(new_state,0) );
          work_queue.push_back( empty_glass(new_state,1) );
          work_queue.push_back( empty_glass(new_state,2) );
        }
        
        work_queue.pop_front();
      }
      ans = volumes.size();
      // for (auto& ans: volumes){
      //   std::cout << "( "<< ans.first << ", " << ans.second << " )" << "\n";
      // }
    }
    return ans;
}


#include "gtest/gtest.h"
//-----------------------------------------------------------------------------
TEST (Glasses, 16_5_3) {
    EXPECT_EQ ( 21, threeGlasses ({16,  5,  3}));
  }
//-----------------------------------------------------------------------------
TEST (Glasses, 1_1_1) {
    EXPECT_EQ (  3, threeGlasses ({ 1,  1,  1}));
  }
//-----------------------------------------------------------------------------
TEST (Glasses, 1_50_100) {
    EXPECT_EQ (151, threeGlasses ({ 1, 50,100}));
  }
//-----------------------------------------------------------------------------
TEST (Glasses, 8_5_3) {
    EXPECT_EQ ( 13, threeGlasses ({ 8,  5,  3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 9_6_3) {
    EXPECT_EQ ( 6, threeGlasses ({ 9,  6,  3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 3_3_3) {
    EXPECT_EQ ( 3, threeGlasses ({  3, 3, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 7_7_7) {
    EXPECT_EQ ( 3, threeGlasses ({  7, 7, 7}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 7_3_3) {
    EXPECT_EQ ( 7, threeGlasses ({  7, 3, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 19_3_3) {
    EXPECT_EQ ( 11, threeGlasses ({  19, 3, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 21_3_3) {
    EXPECT_EQ ( 9, threeGlasses ({  21, 3, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 21_21_3) {
    EXPECT_EQ ( 15, threeGlasses ({  21, 21, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 12_1_1) {
    EXPECT_EQ ( 7, threeGlasses ({  8, 4, 2}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 7_7_3) {
    EXPECT_EQ ( 13, threeGlasses ({  7, 7, 3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 6_2_3) {
    EXPECT_EQ (7, threeGlasses ({  99,66,99}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 47_6_3) {
    EXPECT_EQ (22, threeGlasses ({  47,6,3}));
}
//-----------------------------------------------------------------------------
TEST (Glasses, 41_61_79) {
    EXPECT_EQ (104, threeGlasses ({ 41,61,79}));
}
// -----------------------------------------------------------------------------
TEST (Glasses, 300_1123_389) {
    EXPECT_EQ (200, threeGlasses ({ 98,99,100}));
}

File: source/go/glasses.go
//Author ignitus
//Date   2016-09-19
//File   glasses.go
//Given Three Jugs Find the count of unique volumes obtainable
//by any combination of the following two opertations
//Empty(x)  -> Sets Jugs Volume to 0
//Transfer(x,y) -> Moves a units from X to Y such that Y is full but not overflown.
package main

import (
	"fmt"
	"sort"
)

//-----------------------------------------------------------------------------
type Glass struct {
	volume   int
	capacity int
}
type Glasses []Glass

//-----------------------------------------------------------------------------
func CreateGlass(cap int) Glass {
	return Glass{cap, cap}
}

//-----------------------------------------------------------------------------
func empty_glass(state Glasses, index int) Glasses {
	if len(state) < 3 {
		return Glasses{}
	}
	if state[index].volume != 0 {
		var new_state Glasses
		new_state = append(new_state, state[0])
		new_state = append(new_state, state[1])
		new_state = append(new_state, state[2])
		new_state[index].volume = 0
		if (new_state[0].volume != 0) || (new_state[1].volume != 0) || (new_state[2].volume != 0) {
			// fmt.Printf("empty_glass=%v\n", new_state)
			return new_state
		} else {
			return Glasses{}
		}
	}
	return Glasses{}
}

//-----------------------------------------------------------------------------
func glass_transfer(state Glasses, from int, to int) Glasses {
	if len(state) < 3 {
		return Glasses{}
	}
	if state[from].volume != 0 && state[to].capacity != state[to].volume {
		var new_state Glasses
		new_state = append(new_state, state[0])
		new_state = append(new_state, state[1])
		new_state = append(new_state, state[2])
		vaccent := state[to].capacity - state[to].volume
		avaliable := state[from].volume
		if vaccent >= avaliable {
			new_state[from].volume = 0
			new_state[to].volume += avaliable
		} else {
			avaliable -= vaccent
			new_state[from].volume -= vaccent
			new_state[to].volume += vaccent
		}
		return new_state
	}
	return Glasses{}
}

//-----------------------------------------------------------------------------
func accum_ints(data []int) int {
	ans := int(0)
	for _, e := range data {
		ans += e
	}
	return ans
}

//-----------------------------------------------------------------------------
func accum_glasses(data []Glass) int {
	ans := int(0)
	for _, e := range data {
		ans += e.volume
	}
	return ans
}

//-----------------------------------------------------------------------------
func glass_key(state Glasses) int {
	if len(state) == 3 {
		return state[0].volume<<32 | state[1].volume<<16 | state[2].volume
	} else {
		return 0
	}
}

func threeGlasses(sizes []int) int {
	work_queue := make([]Glasses, 0)
	volumes := make(map[int]int)
	states := make(map[int]int)
	ans := int(1)
	sort.Ints(sizes)
	if sizes[0] == 1 {
		ans = accum_ints(sizes)
	} else {
		volumes[accum_ints(sizes)]++

		work_queue = append(work_queue, Glasses{Glass{0, sizes[0]}, Glass{sizes[1], sizes[1]}, Glass{sizes[2], sizes[2]}})
		work_queue = append(work_queue, Glasses{Glass{sizes[0], sizes[0]}, Glass{0, sizes[1]}, Glass{sizes[2], sizes[2]}})
		work_queue = append(work_queue, Glasses{Glass{sizes[0], sizes[0]}, Glass{sizes[1], sizes[1]}, Glass{0, sizes[2]}})
		states[glass_key(Glasses{Glass{sizes[0], sizes[0]}, Glass{sizes[1], sizes[1]}, Glass{sizes[2], sizes[2]}})]++

		for len(work_queue) != 0 {
			current := work_queue[0]
			if len(current) == 0 {
				work_queue = work_queue[1:]
				continue
			}
			// fmt.Printf("key = %d state=%v  = %d\n", glass_key(current), current, accum_glasses(current))
			volumes[accum_glasses(current)]++
			new_state := glass_transfer(current, 0, 1)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			new_state = glass_transfer(current, 0, 2)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			new_state = glass_transfer(current, 1, 0)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			new_state = glass_transfer(current, 1, 2)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			new_state = glass_transfer(current, 2, 0)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			new_state = glass_transfer(current, 2, 1)
			if states[glass_key(new_state)] == 0 {
				states[glass_key(new_state)]++
				work_queue = append(work_queue, new_state)
				// fmt.Printf("new_state=%v\n", new_state)
				work_queue = append(work_queue, empty_glass(new_state, 0))
				work_queue = append(work_queue, empty_glass(new_state, 1))
				work_queue = append(work_queue, empty_glass(new_state, 2))
			}
			work_queue = work_queue[1:]
			ans = len(volumes)
			// for (auto& ans: volumes){
			//     std::cout << "( "<< ans.first << ", " << ans.second << " )" << "\n";
			// }
		}
	}
	return ans
}

//-----------------------------------------------------------------------------
func EXPECT_EQ(exp int, given int) {
	if exp == given {
		fmt.Printf("Success!\n")
	} else {
		fmt.Printf("Failed:Expected %3d but found %3d\n", exp, given)
	}
}

//-----------------------------------------------------------------------------
func main() {
	EXPECT_EQ(21, threeGlasses([]int{16, 5, 3}))
	EXPECT_EQ(3, threeGlasses([]int{1, 1, 1}))
	EXPECT_EQ(151, threeGlasses([]int{1, 50, 100}))
	EXPECT_EQ(13, threeGlasses([]int{8, 5, 3}))
	EXPECT_EQ(6, threeGlasses([]int{9, 6, 3}))
	EXPECT_EQ(3, threeGlasses([]int{3, 3, 3}))
	EXPECT_EQ(3, threeGlasses([]int{7, 7, 7}))
	EXPECT_EQ(7, threeGlasses([]int{7, 3, 3}))
	EXPECT_EQ(11, threeGlasses([]int{19, 3, 3}))
	EXPECT_EQ(9, threeGlasses([]int{21, 3, 3}))
	EXPECT_EQ(15, threeGlasses([]int{21, 21, 3}))
	EXPECT_EQ(7, threeGlasses([]int{8, 4, 2}))
	EXPECT_EQ(13, threeGlasses([]int{7, 7, 3}))
	EXPECT_EQ(7, threeGlasses([]int{99, 66, 99}))
	EXPECT_EQ(22, threeGlasses([]int{47, 6, 3}))
	EXPECT_EQ(104, threeGlasses([]int{41, 61, 79}))
	EXPECT_EQ(200, threeGlasses([]int{98, 99, 100}))
}
